src/frequenz/lib/notebooks/reporting/metrics/reporting_metrics.py

```
@@ -0,0 +1,304 @@
+
+    Args:
+      prod_excess: Series representing excess production (e.g., kW available for storage).
+      energy_flow_into_battery: Series representing the battery's available charging capacity
+        or power limit at each timestamp.
+
+    Returns:
+      A Series showing the actual production power stored in the battery
+      (elementwise minimum of `prod_excess` and `energy_flow_into_battery`).
+    """
+    prod_excess = prod_excess.astype("float64").clip(lower=0)
+    energy_flow_into_battery = energy_flow_into_battery.astype("float64").clip(lower=0)
+    return pd.concat([prod_excess, energy_flow_into_battery], axis=1).min(axis=1)
+
+
+def grid_feed_in(prod_excess: pd.Series, prod_excess_in_bat: pd.Series) -> pd.Series:
+    """Calculate the portion of excess production fed into the grid.
+
+    Subtracts the amount of excess energy stored in the battery from the total
+    production surplus to determine how much is exported to the grid.
+
+    Args:
+      prod_excess: Series representing the total production surplus
+        (production minus consumption).
+      prod_excess_in_bat: Series representing the portion of that surplus
+        stored in the battery.
+
+    Returns:
+      A Series representing the power or energy fed into the grid.
+    """
+    return (prod_excess - prod_excess_in_bat).clip(lower=0)
+
+
+def prod_self_consumption(asset_prod: pd.Series, prod_excess: pd.Series) -> pd.Series:
+    """Compute the portion of production directly self-consumed.
+
+    Calculates the part of total production that is used locally rather than
+    stored or exported, by subtracting excess production from total production.
+
+    Args:
+      asset_prod: Series representing total production (e.g., kW).
+      prod_excess: Series representing production surplus (portion not directly used).
+
+    Returns:
+      A Series representing self-consumed production (non-negative values only).
+    """
+    return (asset_prod - prod_excess).clip(lower=0)
```

Why clip lower at 0? In fact I would raise an exception or warn if there prod is smaller than excess.

---

Yes, physically it's not possible. Just wanted to make sure that there are no small errors. 
I could add a warning here.